tensorflow/python/kernel_tests/lu_op_test.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests for tensorflow.ops.tf.lu."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+
+from tensorflow.python.client import session
+from tensorflow.python.framework import constant_op
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import linalg_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import variables
+from tensorflow.python.platform import test
+
+
+class LuOpTest(test.TestCase):
+
+  def _verifyLU(self, x):
+    for np_type in [np.float32, np.float64, np.complex64, np.complex128]:
+      if np_type == np.float32 or np_type == np.complex64:
+        tol = 1e-5
+      else:
+        tol = 1e-12
+      if np_type is [np.float32, np.float64]:
+        a = x.real().astype(np_type)
+      else:
+        a = x.astype(np_type)
+      l, u, p, q = linalg_ops.lu(a)
+      pinv = linalg_ops.matrix_inverse(p)
```

Comment:
Inverting a permutation matrix using a general matrix_inverse is terribly inefficient (O(n^3) instead of O(n))! Please use invert_permutation: https://www.tensorflow.org/versions/master/api_docs/python/tf/invert_permutation to invert a permutation, and https://www.tensorflow.org/versions/master/api_docs/python/tf/gather to apply the permutation (or its inverse).

Reply:
Yes. agree. It is ok for the small cases but very bad for the larger ones.

Do not know those APIs before. Will change accordingly. Thanks.

Reply:
@rmlarsen 

Just double check, in the test bed, we write something as follows, right?
```
l, u, p = linalg_ops.lu(a)
pinv = array_ops.invert_permutation(p) #instead of pinv = linalg_ops.matrix_inverse(p)
lu = math_ops.matmul(l, u)
plu = arrays_ops.gather(lu, pinv) #instead of plu = math_ops.matmul(lu, qinv)
```